mlfinlab/backtest_statistics/statistics.py

```diff
@@ -0,0 +1,233 @@
+    """
+    Snippet 14.1, page 197, Derives the timestamps of flattening or flipping
+    trades from a pandas series of target positions. 
+    
+    :param target_positions: (pd.Series) target position series with timestamps as indices
+    :return: (pd.DatetimeIndex) timestamps of trades flattening, flipping and last bet
+    """
+    empty_positions = target_positions[(target_positions==0)].index # Empty positions index
+    previous_positions = target_positions.shift(1) # Timestamps pointing at previous positions
+    previous_positions = previous_positions[(previous_positions!=0)].index # Index of positions where previous one wasn't empty
+    flattening = empty_positions.intersection(previous_positions) # FLATTENING - if previous position was open, but current is empty
+    multiplied_posions = target_positions.iloc[1:] * target_positions.iloc[:-1].values # Multiplies current position with value of next one
+    flips = multiplied_posions[(multiplied_posions<0)].index # FLIPS - if current position has another direction compared to the next
+    res = flattening.union(flips).sort_values()
+    if target_positions.index[-1] not in res:
+        res=res.append(target_positions.index[-1:]) # Appending with last bet
+    return res
+
+
+def average_holding_period(target_positions: pd.Series) -> float:
+    """
+    Snippet 14.2, page 197, Estimates the average holding period (in days) of a strategy, 
+    given a pandas series of target positions using average entry time pairing algorithm.
+    
+    Idea of an algorithm:
+    - entry_time = (previous_time * weight_of_previous_position + 
+                    time_since_beginning_of_trade * increase_in_position ) /
+                    weight_of_current_position
+    - holding_period ['holding_time' = time a position was held, 'weight' = weight of position closed]
+    - res = weighted average time a trade was held
+    
+    :param target_positions: (pd.Series) target position series with timestamps as indices
+    :return: (float) estimated average holding period, NaN if zero or unpredicted
+    """
+    holding_period, entry_time = pd.DataFrame(columns=['holding_time','weight']), 0
+    position_difference = target_positions.diff() 
+    
+    # Time elapsed from the starting time for each position
+    time_difference = (target_positions.index-target_positions.index[0]) / np.timedelta64(1,'D')
+    for i in range(1, target_positions.shape[0]):
+        if float(position_difference.iloc[i] * target_positions.iloc[i-1]) >= 0: # Increased or unchanged position
+            if float(target_positions.iloc[i]) != 0: # And not an empty position
+                entry_time = (entry_time * target_positions.iloc[i-1] + 
+                              time_difference[i] * position_difference.iloc[i]) / target_positions.iloc[i]
+        else: # Decreased
+            if float(target_positions.iloc[i] * target_positions.iloc[i-1]) < 0: # Flip of a position
```

Comment:
Code is a bit long. What is your line length set to? Lets reduce it so that all the code will fit on Github review with out the need to scroll.

Reply:
I didn't have it specified. Will try to fit in 80-100 characters.